fdbclient/DatabaseConfiguration.cpp

```diff
@@ -238,6 +241,15 @@ bool DatabaseConfiguration::isValid() const {
 		}
 	}
 
+	if (encryptionDataAtRestMode != EncryptionDataAtRestMode::DISABLED && tenantMode != TenantMode::REQUIRED) {
+		// FDB data at-rest native encryption defines EncryptionDomain matching TenantBoundaries; tenantMode needs to
+		// 'required' for enabling encryption data at-rest feature
+		TraceEvent("DatabaseConfigurationInvalid")
+		    .detail("EncryptionDataAtRestMode", encryptionDataAtRestMode.toString())
+		    .detail("TenantMode", tenantMode.toString());
+		return false;
```

Comment:
I think this seems ok from a standalone cluster perspective, but we should consider how this fits with a metacluster.

In a metacluster, there is no need to set tenant modes, as they are inferred by metacluster membership. And in fact, the management cluster doesn't behave a like a tenant cluster at all.

I think there are a couple questions we should try to answer about where we're heading, and that might suggest ideas for how we could approach this now.

1. If we are encrypting the system key-space, then presumably we need to do that at creation time? It's possible this state could be configured separately from whether tenants are encrypted and could be treated as a property independent of the metacluster.
2. Are we planning to ever allow configuring encryption per tenant? If so, then it seems like we need to do something like `#1` above and configure tenants and system key-space independently. If not, then in a metacluster we probably don't want encryption to be a cluster property since movement between clusters might get confusing. On the other hand, we have an issue where we write the system key-space before a cluster joins the metacluster.

My initial thought is that the choice to encrypt the system key-space should be a cluster property and the choice to encrypt data should be a metacluster property. Eventually, that encrypted flag could become a tenant property if we wanted that.

Reply:
> If we are encrypting the system key-space, then presumably we need to do that at creation time?

In the current solution, both system and non-system key space; for MetaCluster based on the usecase (need for encyrption is also a consideration, for instance: if system key-space content doesn't need to be encrypted, then may be we don't do it). However, encryption db config better to be `set` only at the `creation` time and then follow `immutable` property afterwards. 

> Are we planning to ever allow configuring encryption per tenant?
That is a good point, maybe we need store `encryption db config` as part of Metacluster config and add it as a property to `TenantInfo` to define if encryption is supported?

Reply:
Summarizing the offline discussion:

Consensus is to make `encryption` independent of cluster `tenant_mode`; for any mutation or KV when getting encrypted, FDB Key will be parsed to extract `tenantId` (encryption domain) information, if valid TenantId exists, then cluster would attempt to use per-tenant encryption keys, otherwise, encryption domain would be mapped to `default domain` per-cluster. 

Thus, I will remove above check, however, `encryption` config will still be restricted to be applied during cluster creation time, once set, the property is immutable.